internal/cluster-object/secret_provider.go

```diff
@@ -0,0 +1,143 @@
+		return nil, err
+	}
+
+	namespaces, err := p.namespaceProvider.All(ctx)
+	if err != nil {
+		p.logger.Error("while fetching namespaces", "error", err)
+		return nil, err
+	}
+
+	nsnames := p.getNamespacesNames(namespaces)
+	if err := p.getAllSecretsWithNamespaceNamePrefix(ctx, secrets, nsnames); err != nil {
+		return nil, err
+	}
+
+	siList, err := p.serviceInstanceProvider.AllWithSecretRef(ctx)
+	if err != nil {
+		p.logger.Error("while fetching service instances with secret ref", "error", err)
+		return nil, err
+	}
+
+	if err := p.getSecretsFromRefInServiceInstances(ctx, siList, secrets); err != nil {
+		return nil, err
+	}
+
+	if len(secrets.Items) == 0 {
+		p.logger.Warn(fmt.Sprintf("no btp operator secrets found"))
+		return nil, err
+	}
+
+	return secrets, err
+}
+
+func (p *SecretProvider) getAllSapBtpServiceOperatorNamedSecrets(ctx context.Context, secrets *corev1.SecretList) error {
+	if err := p.Reader.List(ctx, secrets, client.MatchingFields{"metadata.name": btpServiceOperatorSecretName}); err != nil {
+		p.logger.Error(fmt.Sprintf("failed to fetch all \"%s\" secrets", btpServiceOperatorSecretName), "error", err)
+		return err
+	}
+	return nil
+}
+
+func (p *SecretProvider) getNamespacesNames(namespaces *corev1.NamespaceList) []string {
+	names := make([]string, len(namespaces.Items))
+	for i, ns := range namespaces.Items {
+		names[i] = ns.Name
+	}
+	return names
+}
+
+func (p *SecretProvider) getAllSecretsWithNamespaceNamePrefix(ctx context.Context, secrets *corev1.SecretList, nsnames []string) error {
+	for _, nsname := range nsnames {
+		secret := &corev1.Secret{}
+		secretName := fmt.Sprintf("%s-%s", nsname, btpServiceOperatorSecretName)
+		if err := p.Get(ctx, client.ObjectKey{Namespace: controllers.ChartNamespace, Name: secretName}, secret); err != nil {
+			if k8serrors.IsNotFound(err) {
+				p.logger.Info(fmt.Sprintf("secret \"%s\" not found in \"%s\" namespace", secretName, controllers.ChartNamespace))
+				continue
+			}
+			p.logger.Error(fmt.Sprintf("failed to fetch \"%s\" secret", secretName), "error", err)
+			return err
+		}
+		secrets.Items = append(secrets.Items, *secret)
+	}
+
+	return nil
+}
+
+func (p *SecretProvider) getSecretsFromRefInServiceInstances(ctx context.Context, siList *unstructured.UnstructuredList, secrets *corev1.SecretList) error {
+	for _, item := range siList.Items {
+		secretRef, found, err := unstructured.NestedString(item.Object, "spec", secretRefKey)
+		if err != nil {
+			p.logger.Error(fmt.Sprintf("while traversing \"%s\" unstructured object to find \"%s\" key", item.GetName(), secretRefKey), "error", err)
+			return err
+		} else if !found {
+			p.logger.Warn(fmt.Sprintf("expected secret ref not found in \"%s\" service instance", item.GetName()))
+			continue
+		}
+		secret := &corev1.Secret{}
+		if err := p.Get(ctx, client.ObjectKey{Namespace: controllers.ChartNamespace, Name: secretRef}, secret); err != nil {
```

Comment:
What do you think extracting `Get` to separate method to reduce `getSecretsFromRefInServiceInstances` complexity`?

Reply:
I will make it even better by moving both, finding secret reference key and secret getting, into separate methods